### DISCUSSION TITLE
Speed up pgr_createTopology on large networks?

### DISCUSSION THREAD

**Discussion by kimonkrenz** [pgRouting/pgrouting#2255] [Q&A] [answered]
Is there a way to speed up pgr_createTopology, or modify the number of processed edges per iteration used in pgr_createTopology?

Frequent creations of topologies of larger networks (i.e. 1-8million edges) are very time consuming. 
The algorithm processes 1000 edges per second (approx.), while neither RAM nor CPU are being maxed (not even 5%), resulting in approx. 30min for a 2 million edge network.

```
NOTICE:  PROCESSING:
NOTICE:  pgr_createTopology('road_network', 0.0001, 'geom', 'id', 'source', 'target', rows_where := 'true', clean := t)
NOTICE:  Performing checks, please wait .....
NOTICE:  Creating Topology, Please wait...
NOTICE:  1000 edges processed
NOTICE:  2000 edges processed
NOTICE:  3000 edges processed
NOTICE:  4000 edges processed 
...
```

Nb. the input table is vacuumed, clustered and columns geom, id, source and target are indexed.

Similar questions have been posted at alternative forums (e.g. [stackexchange](https://gis.stackexchange.com/questions/270780/pgr-createtopology-with-large-datasets), or [stackexchange](https://gis.stackexchange.com/questions/69722/why-is-any-pgr-routing-function-taking-forever-based-on-osm-data-in-an-pgrouti)). Unfortunately, existing solutions are for Python which I am not using, or other suggestions don't show any effect.

Set up:
macOS Big Sur 11.6 (20G165), 3.1 GHz Quad-Core Intel Core i7, 16 GB 2133 MHz LPDDR3
PostgreSQL 14.1 on x86_64-apple-darwin20.6.0, compiled by Apple clang version 13.0.0 (clang-1300.0.29.3), 64-bit
POSTGIS="3.1.4 ded6c34" [EXTENSION] PGSQL="140" GEOS="3.10.2-CAPI-1.16.0" SFCGAL="1.4.0" PROJ="7.2.1" GDAL="GDAL 3.4.1, released 2021/12/27" LIBXML="2.9.12" LIBJSON="0.15" LIBPROTOBUF="1.4.0" WAGYU="0.5.0 (Internal)" TOPOLOGY RASTER
pgRouting 3.3.0

**Comment by TimMcCauley** [comment, marked as answer]
@kimonkrenz this is not a full answer to your problem yet but you may experience some speedups using https://docs.pgrouting.org/latest/en/pgr_extractVertices.html instead and afterwards computing contracted edges again https://gis.stackexchange.com/questions/422316/postgis-unsplit-merge-lines-that-have-coincident-endpoints ? What do you think @cvvergara?

**Reply by kimonkrenz** [reply]
Hi @TimMcCauley, many thanks for your suggestions! If I understand you correctly, you recommend to reduce the number of edges within the network via fast vertices extraction and subsequent edge contraction, in order to speed up the topology creation? Unfortunately, the base network I am using is already contracted and cannot be further reduced in size. 

That's why I am more interested in understanding if there are options to improve the performance of pgr_createTopology, as well as to understand which factors influence in general the performance here.

**Reply by TimMcCauley** [reply]
I don't think there are @kimonkrenz - I think this function has not been touched for a while. I tested extractVertices the other day on a grid network, the results can be found [here](https://github.com/pgRouting/pgrouting/issues/2087). Can you share a screenshot of your base network that I understand what you mean by it is already contracted? Are the intersections already geometrically present or do roads have to be split up? What is your input data? Could you not extract the vertices and then remove those again with degree 2?

**Reply by kimonkrenz** [reply]
Hi @TimMcCauley, I think I completely misunderstood your first comment, as I got confused by what role the edge contraction would play (nb my question here was not related to the unsplit merge lines issue on SE). I also didn't realised at first that pgr_extractVertices can generates the topology. It might be worth highlighting this in the documentation. 

Anyway, your suggestion is an excellent recommendation! I managed to create the topology for 2 million edges in under a minute (vs 30min before). I also don't require the tolerance factor, but I guess if needed rounding x/y might generate similar results. Many thanks!

**Reply by TimMcCauley** [reply]
Nice to hear @kimonkrenz - if tolerance is an issue you could also maybe make use of  st_snaptogrid. Happy routing!

**Reply by cvvergara** [reply]
Hi @kimonkrenz sorry for the late reply.
I am glad that you could follow Tim's advice and happy that `pgr_extractVertices` worked for you.

> It might be worth highlighting this in the documentation.

Currently the function is marked as proposed, and the pgRouting's policy is that function marked as proposed move to official on a mayor release.
Actually, my deep soul intention is to rewrite all the topology function, as they modify the tables. and that one was my first attempt, as in my mind it substitutes `pgr_createVerticesTable` where the user is in charge of creating the vertex table.
One of the examples in the documentation is to create a routing topology. (aka the example substitutes `pgr_createTopology`) Need to work out a function that does that task without modifying the edges table because checks about the table take too much time and creating indexes etc take too much time.

So, I was thinking of options to highlight the fact that "pgr_extractVertices can generates the topology"
And my option is to copy the **example 2** in found [here](file:///home/vicky/pgrouting/pgrouting/cvvergara/build/doc/html/en/pgr_extractVertices.html#additional-examples) into `pgr_createTopology`

If you think that solution is the way to highlight it, please (you or @TimMcCauley) open an issue, give a statement of why it is to be considered, like your experiences. (I do not have experience with that function as I test the functions using the sample data of the documentation but I only imagine the big data sets when I write them (aka I work by theory not by practice).
This documentation change can happen on the next micro, 3.3.2 due soon.

And also an issue to move it up to official like this one,
https://github.com/pgRouting/pgrouting/issues/2087
 so that the movement can be done on the next mini 3.4.0 due in September instead of 4.0.0 due in 2023

**Reply by Phyks** [reply]
Hi!

I have a similar problem building topology on a 40M edges graph. `pgr_extractVertices` but in my particular case, I need the `tolerance` parameter.

On a subset of my graph:
* `pgr_createTopology` gives 4787 vertices (with a given tolerance).
* `pgr_extractVertices` gives 4869 vertices.
* Snapping to grid the extracted vertices (with a grid of the same tolerance) gives 4832 vertices (for example `SELECT COUNT(*) FROM (SELECT array_agg(id) FROM vertices_table GROUP BY ST_SnapToGrid(geom, 1e-6)) AS t;
`)
* Using a true distance between vertices gives the same number of vertices as `pgr_createTopology` (for example `SELECT COUNT(*) FROM (SELECT ST_Dump(ST_RemoveRepeatedPoints(ST_Collect(geom), 1e-6)) FROM vertices_table) AS t;`).


`pgr_extractVertices` comes with a huge speed up for me, has anyone already some code / examples to manage the tolerance post-processing?

Thanks!

**Reply by TimMcCauley** [reply]
I'm not sure if I fully understand @Phyks but could you not snap the raw points to a grid beforehand and then run extract vertices?

**Reply by Phyks** [reply]
Thanks a lot for your feedback, that's exactly that.

First of all, I'm not sure why, but I was thinking I needed some post-processing to group nearby `vertices` afterwards. It turns out it is much more efficient to update the SQL query passed to `pgr_extractVertices` to remap the geometries there.

Second, problem with `ST_SnapToGrid` is that you might have rounding errors leading to vertices not being correctly merged. Typical example is a 1D-grid with cell size `0.5` and points `P1(x=1.1)` and `P2(x=1.6)`. Both points are within the same tolerance (0.5) but would not lie within the same grid cell. That is however easily solved if you do not need a strict tolerance (which was my case) by doubling it.

For anyone coming across this discussion and not very familiar with pgRouting advanced methods (as I was), here is a minimal working example for me:

```sql
-- wkb_geometry is in WGS84. I would call pgr_createTopology(..., tolerance:=1e-6), I'm doubling the tolerance here.
SELECT * INTO UNLOGGED vertices_table FROM pgr_extractVertices('SELECT id, ST_SnapToGrid(wkb_geometry, 2e-6) AS geom FROM edges ORDER BY id');

WITH out_going AS (
  SELECT id AS vid, unnest(out_edges) AS eid FROM vertices_table
)
UPDATE edges SET source = vid FROM out_going WHERE id = eid;

WITH in_coming AS (
  SELECT id AS vid, unnest(in_edges) AS eid FROM vertices_table
)
UPDATE edges SET target = vid FROM in_coming WHERE id = eid;
```

and then one can directly use `pgr_dijkstra` on the `edges` table as you would after `pgr_createTopology`.


I am handling a graph with 40M edges. `pgr_createTopology` would take 20-30 hours and have to run in batches to avoid OOM. Switching to `pgr_extractVertices`, I am now able to build the topology in about 1 hour.

Many thanks!

**Reply by TimMcCauley** [reply]
That's indeed great to hear and thank you for the recipe! The `pgr_extractVertices` approach will add vertices everywhere, i.e. also on those nodes that aren't decision points in the topology. It would be interesting to see if somebody has managed to remove those from the topology and only use them for the shape. Something similar to this [approach](https://gis.stackexchange.com/questions/422316/postgis-unsplit-merge-lines-that-have-coincident-endpoints). This is the advantage of `pgr_createTopology` as it only keeps the relevant information.  Maybe one reason why it is so much slower? I also believe the traversal would be faster on removing superfluous vertices.

**Reply by Phyks** [reply]
On my particular use case, I am dealing with a tree-like graph topology. I don't expect too many issues regarding the extra nodes, and worst case scenario, basic filtering on bbox should do the trick.

>  Maybe one reason why it is so much slower?

I might be missing something, but contracting edges as in [your example](https://gis.stackexchange.com/questions/422316/postgis-unsplit-merge-lines-that-have-coincident-endpoints) should boil down to perform a DFS and merging two consecutive edges if they are the only ones connected by a given vertices?

With the appropriate indices in place and once the topology has been created, I would expect this to require no more than one sequential scan of each of the involved tables (`edges`  and `vertices`) and would not expect it to account for the runtime difference between `pgr_createTopology` and `pgr_extractVertices`?

**Reply by TimMcCauley** [reply]
That makes sense to me. This scenario is fairly common for OSM use cases in which roads consist of multiple support points. @cvvergara @dkastl do you think this mix of functions could in sequence make up a signature of a "create topology version 2"? ☀️

**Reply by Phyks** [reply]
For my particular use case, I need to run a `pgr_connectedComponents`  analysis on my graph (40M edges). Trying to run it directly leads into a malloc error:

> invalid memory alloc request size 1080000000

So I though I could look into the contraction part to downsize my graph, and tried to run `pgr_contraction` as per [your example](https://gis.stackexchange.com/questions/422316/postgis-unsplit-merge-lines-that-have-coincident-endpoints). I get the exact same malloc issue with `pgr_contraction`.

This is not so much a blocking issue when using `pgr_dijkstra` since, although my graph is large, routing is very local and I can easily filter out geometries with a bounding box to reduce number of considered edges.

I'm thinking about tiling the runs of `pgr_contraction`  or `pgr_connectedComponents` (by gridding with tiles of a few degrees latitude/longitude for instance) and then merging the results. But maybe someone around would have a better idea?


I'm also quite surprised by the malloc error message. I'm not sure about the unit of the number displayed, but if I'm counting right, this should not be much more than 1GB (while my laptop certainly does not have plenty of RAM, it should have at least 10GB free).

Thanks!

**Reply by Phyks** [reply]
For my particular use case, I need to run a `pgr_connectedComponents`  analysis on my graph (40M edges). Trying to run it directly leads into a malloc error:

> invalid memory alloc request size 1080000000

So I though I could look into the contraction part to downsize my graph, and tried to run `pgr_contraction` as per [your example](https://gis.stackexchange.com/questions/422316/postgis-unsplit-merge-lines-that-have-coincident-endpoints). I get the exact same malloc issue with `pgr_contraction`.

This is not so much a blocking issue when using `pgr_dijkstra` since, although my graph is large, routing is very local and I can easily filter out geometries with a bounding box to reduce number of considered edges.

I'm thinking about tiling the runs of `pgr_contraction`  or `pgr_connectedComponents` (by gridding with tiles of a few degrees latitude/longitude for instance) and then merging the results. But maybe someone around would have a better idea?


I'm also quite surprised by the malloc error message. I'm not sure about the unit of the number displayed, but if I'm counting right, this should not be much more than 1GB (while my laptop certainly does not have plenty of RAM, it should have at least 10GB free).

Thanks!

EDIT: My malloc issue looks like https://github.com/pgRouting/pgrouting/issues/291 but I'm not seeing much solutions there :/

**Reply by TimMcCauley** [reply]
This unfortunately leaves my realm of knowledge, the only thing I can think of is tinkering around with the memory settings in `postgresql.conf` - you can have a look at pgtune for some ideas.

**Reply by Phyks** [reply]
I have the exact same issue with a default `postgresql.conf` or one from `pgtune`.

My current `postgresql.conf` looks like:

```
#------------------------------------------------------------------------------
# CUSTOMIZED OPTIONS
#------------------------------------------------------------------------------

# Add settings for extensions here

# https://pgtune.leopard.in.ua/
# DB Version: 14
# OS Type: linux
# DB Type: mixed
# Total Memory (RAM): 12 GB
# CPUs num: 4
# Connections num: 50
# Data Storage: ssd

max_connections = 50
shared_buffers = 3GB
effective_cache_size = 9GB
maintenance_work_mem = 768MB
checkpoint_completion_target = 0.9
wal_buffers = 16MB
default_statistics_target = 100
random_page_cost = 1.1
effective_io_concurrency = 200
work_mem = 15728kB
min_wal_size = 1GB
max_wal_size = 4GB
max_worker_processes = 4
max_parallel_workers_per_gather = 2
max_parallel_workers = 4
max_parallel_maintenance_workers = 2
```

And the error message:

```
db=# (SELECT component, node FROM pgr_connectedComponents(
        'SELECT id, source, target, cost  FROM edges'
));
ERREUR:  invalid memory alloc request size 1080000000
CONTEXTE : fonction SQL « pgr_connectedcomponents », instruction 1
```

I'll post on the pgRouting ML, it might get more visibility there to finalize an end to end recipe for handling large graphs with `pgr_extractVertices`.

**Reply by TimMcCauley** [reply]
@cvvergara 👋🏽 have you seen something similar on larger graphs?
